### ISSUE TITLE
Getting exception if creating a record with an invalid relation

### ISSUE THREAD

**Issue by efatsi** [neo4jrb/neo4j#1314]
Curious if this is expected behavior. When creating a new resource, if you pass in an invalid value for a related resource, the code blows up when it then tries to reference that relation. I've run into this problem in trying to validate the presence of a relationship. Passing in `"invalid-uuid"`, or `nil` raises an exception.
### Code example (inline, gist, or repo)

``` ruby
class Book
  include Neo4j::ActiveNode
  property :title, type: String

  has_one :in, :author, type: :WROTE, model_class: :User

  validates :title, :author, presence: true
end
```

``` bash
irb(main):001:0> b = Book.new(author: "invalid-uuid")
=> #<Book uuid: nil, title: nil>
irb(main):002:0> b.valid?
 Book#author 6ms 
  MATCH (book)
  WHERE (ID(book) IS NULL)
  MATCH (book)<-[rel1:`WROTE`]-(result_author:`User`)
  RETURN result_author 
 User 3ms 
  MATCH (n:`User`)
  WHERE (n.uuid = {n_uuid})
  RETURN n
  ORDER BY n.uuid
  LIMIT {limit_1} | {:n_uuid=>"invalid-uuid", :limit_1=>1}
Neo4j::ActiveNode::Labels::RecordNotFound: Couldn't find User with 'uuid'=invalid-uuid
```
### Runtime information:

Neo4j database version:
`neo4j` gem version: 7.2.3
`neo4j-core` gem version: 6.1.6

**Comment by cheerfulstoic** [comment]
This is what I would expect.  Is there something that you would expect from a different ORM?

As a side note, I don't know offhand how ActiveRecord would handle passing in the ID for the `author` association, but I *think* you need to specify `author_id` (generally because that's the actual name of the column in the table where you would be storing the value).  With `ActiveNode` associations support this syntax as well (`author_id` / `author_id=` for `has_one` associations and `author_ids` / `author_ids=` for `has_many` associations)

**Comment by efatsi** [comment]
My expectation would be that if an invalid value is passed in, then the ORM would return `nil` for a subsequent check on the association instead of raising `RecordNotFound`.

The primary use case for this I see is when `nil` is passed in (ie: someone doesn't select an Author from a select dropdown and submits the "create Book" form).

**Comment by cheerfulstoic** [comment]
`nil` is different, is that returning `RecordNotFound`?  I would expect that `nil` would remove the relationshp

**Comment by anamba** [comment]
I'm running into this issue when trying to add a simple presence validation to an association. (I am new to neo4j, maybe this is some kind of antipattern? Let me know if it is, I googled and didn't turn up anything on this either way.)

I was able to use `validate :custom_method_name` but anything else (including a custom ActiveModel::EachValidator) doesn't work because ActiveModel helpfully tries to get me the value of the association, which then raises the RecordNotFound exception.

So right now I am adding ~7 lines of boilerplate code for each assocation I want to validate (calling the association, catching the exception, adding a message to record.errors), which doesn't seem particularly Rails-y.

**Comment by cheerfulstoic** [comment]
Hey @anamba !

I don't think I completely understand.  Could you share some sample code (both what you first tried as well as the ~7 lines of boilerplate) so I could have a better idea of what the trouble is?

**Comment by anamba** [comment]
Sure, here's what I'm seeing. ActiveNode doesn't have a belongs_to association but that is basically what I'm doing here, so I wanted to replicate the default ActiveRecord behavior of ensuring that the belongs_to association is connected properly on save.

First attempt:
```
  validates :vendor, presence: true
```
raises this exception on when I try to save without a vendor: `Neo4j::ActiveNode::Labels::RecordNotFound (Couldn't find Vendor with 'uuid'=):`

So next, I tried simply rescuing that exception:
```
  validate :vendor_presence
  
  def vendor_presence
    begin
      vendor
    rescue Neo4j::ActiveNode::Labels::RecordNotFound => e
      errors.add(:vendor, 'is required')
      return false
    end
  end
```
That worked, but I don't want to have to copy and paste that into every model, so I tried this:
```
class AssociationPresenceValidator < ActiveModel::EachValidator
  def validate_each(record, attribute, value)
    logger.debug "AssociationPresenceValidator called: (record = #{record.inspect}, attribute = #{attribute.inspect}, value = #{value.inspect})"
    begin
      record.send(attribute)
    rescue Neo4j::ActiveNode::Labels::RecordNotFound => e
      record.errors.add(attribute, options[:message] || 'is required')
    end
  end
end
```
and in the model file:
```
validates :vendor, association_presence: true
```
But the debug doesn't even output (actually looking at it now, I think it would have raised an exception, should have been Rails.logger.debug) because it dies when the framework tries to pass it the value argument. Same exception:
`Neo4j::ActiveNode::Labels::RecordNotFound (Couldn't find Vendor with 'uuid'=):`

I suppose if this behavior can't be changed, I could also try creating a concern, but it would be nice if it could just return nil instead of raising an exception.

**Comment by cheerfulstoic** [comment]
Regarding `belongs_to`, the reason the gem doesn't have that is because in `ActiveRecord` the `has_one` and `belongs_to` are essentially the same association but for different cases in the DB (`belongs_to` is when you have a foreign key referring to another model while `has_one` is when another model is referring to you).  In Neo4j everything is a relationship, and so there is only a need for a `has_one` association.

But I think see what you're doing now.  How is `vendor` defined, by the way?  It is strange that you would get the `Couldn't find Vendor with 'uuid'=):` error with just a pure association because it doesn't search for the node by ID but rather it traverses the relationship

**Comment by anamba** [comment]
Yes, I understand why belongs_to is not needed, just thought that mentioning it would be an easy way to explain what I was doing and why.

I thought the uuid error was odd, too. My specs pass (`it { should validate_presence_of :vendor }`) but when I try to create a record in my web browser without specifying a vendor, that's when I get that error.

To keep things simple, the Rails code involved is straight from the scaffold generator. Here's the relevant controller action:
```
  def create
    @datacenter = Datacenter.new(datacenter_params)
    if @datacenter.save
      redirect_to @datacenter, notice: 'Datacenter was successfully created.'
    else
      render :new
    end
  end
```

**Comment by anamba** [comment]
Aha... after contemplating the error message itself for a while more, I finally figured out where it was coming from. It definitely wasn't neo4jrb's fault, but I'll post it here in case someone else makes the same mistake I did.

At last, the culprit:
```
    def datacenter_params
      params.require(:datacenter).permit(:name, :vendor_id)
    end
```
A while back, I put :vendor_id in the list of permitted parameters, hoping that would "just work" the way it does with ActiveRecord. It didn't, of course, because that's not how relations work with neo4j, so I later handled it more like this:
```
      if v = Vendor.find_by_id(params[:datacenter][:vendor_id])
        @datacenter.vendor = v
      end
```
However, I forgot to remove that :vendor_id from the list of permitted parameters, and when that blank vendor_id gets assigned, that's when the uuid exception gets thrown. Mystery solved!

Sorry for wasting your time @cheerfulstoic, but thanks for helping me work through this. Your questions did ultimately help me find the issue. And hopefully this helps someone else who comes across this odd exception.

**Comment by cheerfulstoic** [comment]
FYI, I believe that `it { should validate_presence_of :vendor }` simply looks at the validation definitions on the model class and doesn't try to exercise them.

I see why it would do that now!  Really it probably is the gem's fault a bit.  I think that when the code raises the exception it should use `.inspect` on the value, so the error would be `Couldn't find Vendor with 'uuid'=nil`, which I think might help to make things more clear.  I just made that fix and released it as version 9.1.1.  Thanks!
